compiler/optimizer/LoopVersioner.cpp

```diff
@@ -4905,8 +4909,29 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
          succNode->removePredecessor(succEdge);
          subNode->removeSuccessor(succEdge);
 
+         TR::CFGEdge *toInvariantEdge = NULL;
+
          for (auto changedSuccEdge = succNode->getSuccessors().begin(); changedSuccEdge != succNode->getSuccessors().end(); ++changedSuccEdge)
+            {
+            TR::CFGEdge *currEdge = *changedSuccEdge;
+
+            // Keep track of any edge from the original loop to the invariant
+            // block - it must be discarded from the list of successors of the
+            // new proper region in the context of the parent structure, as a
+            // subgraph node must not have an edge to itself
+            if (currEdge->getTo() == properNode)
+               {
+               toInvariantEdge = currEdge;
+               }
+
             (*changedSuccEdge)->setFrom(properNode);
+            }
+
+         if (toInvariantEdge)
+            {
+            properNode->removeSuccessor(toInvariantEdge);
```

Comment:
Should this be ```succNode``` instead of ```properNode``` for the ```removeSuccessor``` call ?

Reply:
@vijaysun-omr Thanks for reviewing, Vijay.  The reason that this is removing `invariantEdge` as a successor edge of `properNode` is that `properNode` was just made into the successor in the call to `TR::CFGEdge::setFrom` in the `for` loop above.

I guess another possibility would have been to avoid changing the successor at all in that case, as follows, eliminating the need for the subsequent call to `removeSuccessor`:

```
         for (auto changedSuccEdge = succNode->getSuccessors().begin(); changedSuccEdge != succNode->getSuccessors().end(); ++changedSuccEdge)
            {
            TR::CFGEdge *currEdge = *changedSuccEdge;

            // Keep track of any edge from the original loop to the invariant
            // block - it must be discarded from the list of successors of the
            // new proper region in the context of the parent structure, as a
            // subgraph node must not have an edge to itself
            if (currEdge->getTo() != properNode)
               {
               (*changedSuccEdge)->setFrom(properNode);
               }
            else
               {
               toInvariantEdge = currEdge;
               }
            }

         if (toInvariantEdge)
            {
            properNode->removePredecessor(toInvariantEdge);
            }
```

In either case, I think there's no need to call `removeSuccessor` or `removePredecessor` on `succNode`, as it's completely removed from the `parentStructure` just after all its exit and exception edges are transferred to `properNode`.

Would you prefer the current approach or the revision I propose above or a different alternative?

Reply:
The loop just above has done `toInvariantEdge->setFrom(properNode)`, which put `toInvariantEdge` into `properNode->_successors`, so it has to be removed from there. Note that `setFrom()` doesn't remove the edge from the successors of the original source, so `succNode->_successors` is unchanged, i.e. it includes all of the edges that are now in `properNode->_successors`. That's fine because the successors are ignored by `removeSubNode(succNode)` just below, and they aren't used in the interim.

Reply:
I'm just testing on a slight revision to the segment of code that we were discussing here.  I hope that will make it a bit more clear what is happening.

Reply:
@vijaysun-omr Vijay, I've tried to simplify the change in this region to make it a bit more clear what is happening in commit https://github.com/eclipse/omr/pull/3745/commits/d04eae10aa140bb244a6bdbe0c56de897c212860.  If you're OK with it, I will squash that commit down.